LLVM -Og leads to invalid wasm (bad drop/fallthru)

We have issues when compiling our source code with some debug information. If we build in Release mode without additional debugging flags, everything is working just fine.

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 4.0.13 (2659582941bef14008476903f48941909db1b196)

**Error log**
```
[wasm-validator error in function __original_main] unexpected false: non-final block elements returning a value must be dropped, on 
(block (result i32)
 (call $__pthread_self_internal)
 (i64.const 2)
 (i32.const 0)
)
(on index 0:
0x40d0c360310
), type: i64
[wasm-validator error in function __original_main] unexpected false: non-final block elements returning a value must be dropped, on 
(block (result i32)
 (call $__pthread_self_internal)
 (i64.const 2)
 (i32.const 0)
)
(on index 1:
0x40d0c360358
), type: i64
Fatal: error validating input
em++: error: '/emsdk/upstream/bin/wasm-opt --print-function-map main.wasm --mvp-features --enable-bulk-memory --enable-bulk-memory-opt --enable-call-indirect-overlong --enable-memory64 --enable-multivalue --enable-mutable-globals --enable-nontrapping-float-to-int --enable-reference-types --enable-sign-ext' failed (returned 1)
```

**Simple reproducer**
`mkdir build && cd build && emcmake cmake .. -DCMAKE_BUILD_TYPE=Release && cmake --build .`

```cpp
// cat main.cpp
#include <emscripten/bind.h>

int main() {
    emscripten::val a;
    return a.isNull();
}
```

```CMake
# cat CMakeLists.txt
project(reproducer LANGUAGES CXX)
 
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
option(ENABLE_WASM64 OFF)
 
add_executable(main main.cpp)
target_compile_options(main PRIVATE -Og -sMEMORY64=1)
target_link_options(main PRIVATE -Og --bind --emit-symbol-map -sMEMORY64=1)
```

The whole command is as follows:

```
/emsdk/upstream/emscripten/em++ -O3 -DNDEBUG -Og --bind --emit-symbol-map -sMEMORY64=1 CMakeFiles/main.dir/main.o -o main.js   && :
[wasm-validator error in function __original_main] unexpected false: non-final block elements returning a value must be dropped, on 
(block (result i32)
 (call $__pthread_self_internal)
 (i64.const 2)
 (i32.const 0)
)
(on index 0:
0x3b028360310
), type: i64
[wasm-validator error in function __original_main] unexpected false: non-final block elements returning a value must be dropped, on 
(block (result i32)
 (call $__pthread_self_internal)
 (i64.const 2)
 (i32.const 0)
)
(on index 1:
0x3b028360358
), type: i64
Fatal: error validating input
em++: error: '/emsdk/upstream/bin/wasm-opt --print-function-map main.wasm --mvp-features --enable-bulk-memory --enable-bulk-memory-opt --enable-call-indirect-overlong --enable-memory64 --enable-multivalue --enable-mutable-globals --enable-nontrapping-float-to-int --enable-reference-types --enable-sign-ext' failed (returned 1)
ninja: build stopped: subcommand failed.
```

I did test this on the recent release of Emscripten (4.0.15) and while if did not crash during the compilation, we hit an abort when we tried to load the compiled module on the web with a similar message about value being dropped.

--- a/llvm/lib/Target/WebAssembly/WebAssemblyExplicitLocals.cpp
+++ b/llvm/lib/Target/WebAssembly/WebAssemblyExplicitLocals.cpp
@@ -219,7 +219,7 @@ static MachineInstr *findStartOfTree(MachineOperand &MO,
 // FAKE_USEs are no-ops, so remove them here so that the values used by them
 // will be correctly dropped later.
 static void removeFakeUses(MachineFunction &MF) {
-  SmallVector<MachineInstr*> ToDelete;
+  SmallVector<MachineInstr *> ToDelete;
   for (auto &MBB : MF)
     for (auto &MI : MBB)
       if (MI.isFakeUse())

--- a/llvm/lib/Target/WebAssembly/WebAssemblyExplicitLocals.cpp
+++ b/llvm/lib/Target/WebAssembly/WebAssemblyExplicitLocals.cpp
@@ -216,6 +216,18 @@ static MachineInstr *findStartOfTree(MachineOperand &MO,
   return Def;
 }
 
+// FAKE_USEs are no-ops, so remove them here so that the values used by them
+// will be correctly dropped later.
+static void removeFakeUses(MachineFunction &MF) {
+  SmallVector<MachineInstr*> ToDelete;
+  for (auto &MBB : MF)
+    for (auto &MI : MBB)
+      if (MI.isFakeUse())
+        ToDelete.push_back(&MI);
+  for (auto *MI : ToDelete)
+    MI->eraseFromParent();
+}
+
 bool WebAssemblyExplicitLocals::runOnMachineFunction(MachineFunction &MF) {
   LLVM_DEBUG(dbgs() << "********** Make Locals Explicit **********\n"
                        "********** Function: "
@@ -225,6 +237,8 @@ bool WebAssemblyExplicitLocals::runOnMachineFunction(MachineFunction &MF) {
   MachineRegisterInfo &MRI = MF.getRegInfo();
   WebAssemblyFunctionInfo &MFI = *MF.getInfo<WebAssemblyFunctionInfo>();
   const auto *TII = MF.getSubtarget<WebAssemblySubtarget>().getInstrInfo();
+
+  removeFakeUses(MF);
 
   // Map non-stackified virtual registers to their local ids.
   DenseMap<unsigned, unsigned> Reg2Local;

--- a/llvm/test/CodeGen/WebAssembly/fake-use.ll
+++ b/llvm/test/CodeGen/WebAssembly/fake-use.ll
@@ -1,0 +1,15 @@
+; RUN: llc < %s | llvm-mc -triple=wasm32-unknown-unknown
+
+target triple = "wasm32-unknown-unknown"
+
+define void @fake_use_test() {
+  %t = call i32 @foo()
+  tail call void (...) @llvm.fake.use(i32 %t)
+  ret void
+}
+
+declare void @foo()
+; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: readwrite)
+declare void @llvm.fake.use(...) #0
+
+attributes #0 = { mustprogress nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: readwrite) }

